File: exercism/clojure/nucleotide-count/src/nucleotide_count.clj
(ns nucleotide-count)

(defn count-of-nucleotide-in-strand [nucleotide strand]
  (if (and (every? (partial contains? #{\A, \T , \C , \G }) strand)
           (contains? #{\A, \T , \C , \G } nucleotide))
    (->> strand (filter (partial = nucleotide)) (count))
    (throw (IllegalArgumentException.))
    )

)


(defn nucleotide-counts [strand]
  {
   \A (count-of-nucleotide-in-strand \A strand),
   \T (count-of-nucleotide-in-strand \T strand),
   \G (count-of-nucleotide-in-strand \G strand),
   \C (count-of-nucleotide-in-strand \C strand)
   }
)

File: exercism/julia/nucleotide-count/nucleotide-count.jl
"""
    count_nucleotides(strand)

The count of each nucleotide within `strand` as a dictionary.

Invalid strands raise a `DomainError`.

"""
function count_nucleotides(strand)
    freqs = Dict('A' => 0, 'C' => 0, 'G' => 0, 'T' => 0)
    for n in strand
        try
            freqs[n] += 1
        catch err
            throw(DomainError(n, "$(n) is not a nucleotide"))
        end
    end
    return freqs
end
